### PR TITLE
Add Poisson node

### PR DESCRIPTION
**Pull request by albertpod**
This PR addresses issue #129

### CODE REVIEW

File: src/rules/poisson/out.jl [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export rule
+
+@rule Poisson(:out, Marginalisation) (m_l::PointMass,) = Poisson(mean(m_l))
+
+@rule Poisson(:out, Marginalisation) (q_l::GammaDistributionsFamily,) = Poisson((shape(q_l) - 1 // 2) / rate(q_l))
```
Comment (bartvanerp) [inline src/rules/poisson/out.jl:5]: I end up with a digamma function here: `lambda = exp(digamma(alpha))/beta`. Also saw that there are some approximations to this term: https://en.wikipedia.org/wiki/Digamma_function#Computation_and_approximation. The question is whether we always want to use this approximation.

File: src/nodes/poisson.jl [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        end
+        return s
+    else
+        # Try `BigFloat`
+        return convert(T, __approximate_powersum(BigFloat, l, 150))
+    end
+end
+
+@average_energy Poisson (q_out::Any, q_l::Any) =
+    mean(q_l) - mean(q_out) * mean(log, q_l) + exp(-mean(q_out)) * __approximate_powersum(Float64, mean(q_out))
```
Comment (bartvanerp) [inline src/nodes/poisson.jl:29]: Derivation seems correct. I am not really up to speed of the approximation of `E_q(x)[ln(x!)]`, but could we also get away with the (perhaps simpler) Stirling's approximation? Furthermore I think we want to add some specialization for this average energy function, such that we do not always compute this approximation.

Comment (albertpod) [reply src/nodes/poisson.jl:29]: Will it actually help?

Comment (bvdmitri) [reply src/nodes/poisson.jl:29]: I think you still need to approximate this power sum even if `q_out` is a `PointMass`. We've spent quite some time yesterday to figure out how we can further simplify this expression. The computation is super unstable for large values. Distributions.jl has an interesting approximation here: https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/discrete/poisson.jl#L68, but I'm not sure how to incorporate it in our case (it should be possible, maybe I will spent more time on this). They do compute a slightly different quantity though. Any cool ideas are welcome!

Comment (bvdmitri) [reply src/nodes/poisson.jl:29]: Ah, well, I understood you wrong. You are correct @bartvanerp if `Poisson` used for observations. The original approximation, however, assumes Poisson as a prior, not as a likelihood for your observations. So we need it anyway I believe. But it seems reasonable to add an extra dispatch method where `q_out` is a `PointMass`, it simplifies the computation indeed.

**Comment by albertpod** [comment]
Please look up the fixes

**Review by bartvanerp — APPROVED** [review]

**Review by bvdmitri — APPROVED** [review]

**Pull request merged**